File: q0781/q0781.go
func numRabbits(answers []int) int {
    groupsBySizes := make(map[int]int)
    for _, ans := range answers {
        // Get group size
        size := ans + 1
        count, exists := groupsBySizes[size]
        if exists {
            groupsBySizes[size] = count + 1
        } else {
            groupsBySizes[size] = 1
        }
    }

    totalRabbits := 0
    for size, count := range groupsBySizes {
        numOfGroups := count / size
        if count % size != 0 {
            numOfGroups++
        }
        totalRabbits += (size * numOfGroups)
    }
    return totalRabbits
}

File: q0781/q0781.java
class Solution {
    public int numRabbits(int[] answers) {
        Map<Integer, Integer> groupsBySizes = new HashMap<>();
        for (int ans : answers) {
            // Get the size of the group
            int size = ans + 1; 
            groupsBySizes.put(size, groupsBySizes.getOrDefault(size, 0) + 1);
        }

        int totalRabbits = 0;
        for (Map.Entry<Integer, Integer> entry : groupsBySizes.entrySet()) {
            int size = entry.getKey();
            int count = entry.getValue();
            // Calculate the number of full groups
            int numFullGroups = (count / size) + (count % size != 0 ? 1 : 0);
            totalRabbits += (size * numFullGroups);
        }
        return totalRabbits;
    }
}
